File: src/main/java/kitchenpos/products/tobe/domain/Product.java
package kitchenpos.products.tobe.domain;

import jakarta.persistence.AttributeOverride;
import jakarta.persistence.Column;
import jakarta.persistence.Embedded;
import jakarta.persistence.EmbeddedId;
import jakarta.persistence.Entity;
import jakarta.persistence.Table;
import java.math.BigDecimal;
import java.util.UUID;
import kitchenpos.shared.client.PurgomalumClient;
import kitchenpos.shared.domain.ProductPrice;

@Table(name = "product")
@Entity(name = "TobeProduct")
public class Product {

    @EmbeddedId
    @AttributeOverride(name = "value", column = @Column(name = "id"))
    private ProductId id;

    @Embedded
    @AttributeOverride(name = "value", column = @Column(name = "name"))
    private ProductName name;

    @Embedded
    @AttributeOverride(name = "value", column = @Column(name = "price"))
    private ProductPrice price;

    protected Product() {
    }

    public Product(final UUID id, final String name, final PurgomalumClient purgomalumClient,
        final BigDecimal price) {
        this.id = new ProductId(id);
        this.name = new ProductName(name, purgomalumClient);
        this.price = new ProductPrice(price);
    }

    public UUID getId() {
        return id.getValue();
    }

    public String getName() {
        return name.getValue();
    }

    public BigDecimal getPrice() {
        return price.getValue();
    }

    public void changePrice(BigDecimal price) {
        this.price = new ProductPrice(price);
    }
}

File: src/main/java/kitchenpos/products/tobe/domain/Product.kt
package kitchenpos.products.tobe.domain

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.Id
import jakarta.persistence.Table
import kitchenpos.common.Price
import java.util.*

@Table(name = "product")
@Entity
class Product(name: ProductName, price: Price) {
    @Id
    @Column(name = "id", columnDefinition = "binary(16)")
    val id: UUID = UUID.randomUUID()

    @Column(name = "name", nullable = false)
    val name: ProductName = name

    @Column(name = "price", nullable = false)
    var price: Price = price
        private set

    fun changePrice(price: Price) {
        this.price = price
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Product

        return id == other.id
    }

    override fun hashCode(): Int {
        return id.hashCode()
    }

}
